# **Laporan Praktikum**
## **Kecerdasan Komputasional (Probability and Uncertainty)**


### **Identitas Praktikan**
| **Nama**                | **NRP**        | **Kelas**      | **Jurusan**                   |
|--------------------------|----------------|----------------|-------------------------------|
| Wayan Raditya Putra      | 5054241029     | N | Rekayasa Kecerdasan Artifisial      |

### **Dosen Pengampu**
| **Nama Dosen**          |
|--------------------------|
| Imam Mustafa Kamal, S.ST, Ph.D |


### **Judul Praktikum**
Representasi Ketidakpastian Menggunakan Probabilitas dan Distribusi

### **Tujuan**
Mempelajari bagaimana probabilitas digunakan untuk merepresentasikan ketidakpastian dalam sistem kecerdasan komputasional, serta memahami penerapan konsep dasar seperti variabel acak, distribusi peluang, dan inferensi probabilistik.




### **Inisialisasi dan Import Library**

Pada bagian ini dilakukan proses *import* terhadap beberapa modul yang digunakan dalam praktikum.  
Baris kode berikut:



In [6]:
from probability import *
from utils import print_table
from notebook import psource, pseudocode, heatmap

## PROBABILITY DISTRIBUTION

In [7]:
psource(ProbDist)


Kelas ProbDist

Bagian ini mendefinisikan kelas ProbDist yang berfungsi untuk merepresentasikan distribusi probabilitas diskret (discrete probability distribution) dari sebuah variabel acak.

Distribusi probabilitas diskret menggambarkan kumpulan nilai-nilai yang mungkin dari suatu variabel acak beserta peluang kemunculannya, di mana total seluruh peluang bernilai 1.

Penjelasan fungsi dan atribut utama:

1. ****init**(self, var_name='?', freq=None)**
   Konstruktor kelas yang menginisialisasi objek ProbDist.
   Parameter var_name adalah nama variabel acak (misalnya Flip, X, dan sebagainya).
   Parameter freq bersifat opsional dan berisi pasangan nilai–frekuensi dalam bentuk dictionary. Jika diberikan, distribusi akan otomatis dinormalisasi agar total probabilitasnya menjadi 1.

2. ****getitem**(self, val)**
   Mengembalikan nilai probabilitas untuk suatu value tertentu, misalnya P['H']. Jika nilai belum pernah ditentukan, akan dikembalikan 0.

3. ****setitem**(self, val, p)**
   Menetapkan probabilitas p untuk sebuah nilai val. Jika val belum ada dalam daftar nilai, otomatis ditambahkan ke atribut values.

4. **normalize(self)**
   Melakukan normalisasi, yaitu memastikan total semua probabilitas berjumlah 1.
   Hal ini penting karena dalam teori probabilitas, ∑P(x) = 1 merupakan syarat distribusi valid.
   Jika total awal bukan 1, semua nilai akan dibagi dengan totalnya agar memenuhi syarat tersebut.

5. **show_approx(self, numfmt='{:.3g}')**
   Menampilkan isi distribusi dalam format teks yang rapi dan dibulatkan (rounded), berguna untuk dokumentasi atau pengujian.

6. ****repr**(self)**
   Menampilkan representasi objek dalam format P(var_name), sehingga memudahkan pembacaan saat objek dipanggil di interpreter.

Kesimpulannya, kelas ProbDist berperan sebagai pondasi dasar dalam representasi probabilistik di notebook ini. Kelas ini digunakan untuk menyimpan, mengakses, dan menormalkan nilai peluang secara terstruktur serta mudah dioperasikan pada eksperimen berikutnya.


In [8]:
p = ProbDist('Flip')
p['H'], p['T'] = 0.25, 0.75
p['T']

0.75

Potongan kode di atas digunakan untuk membuat dan menguji objek dari kelas ProbDist. Objek bernama p dideklarasikan dengan variabel acak “Flip” yang merepresentasikan percobaan lemparan koin. Selanjutnya ditetapkan nilai probabilitas untuk dua kemungkinan hasil, yaitu sisi “H” (Head) sebesar 0.25 dan sisi “T” (Tail) sebesar 0.75. Baris terakhir p['T'] berfungsi untuk menampilkan nilai probabilitas dari sisi “T”, yaitu 0.75. Dengan demikian, kode ini menunjukkan cara mendefinisikan distribusi probabilitas sederhana dan mengakses nilai peluang dari suatu variabel acak.


In [9]:
p = ProbDist(freq={'low': 125, 'medium': 375, 'high': 500})
p.var_name

'?'

Potongan kode ini membuat objek distribusi probabilitas bernama p menggunakan parameter freq yang berisi pasangan nilai–frekuensi, yaitu ‘low’ sebesar 125, ‘medium’ sebesar 375, dan ‘high’ sebesar 500. Pada saat objek dibuat, konstruktor kelas ProbDist secara otomatis akan menormalkan frekuensi tersebut sehingga total probabilitasnya menjadi 1. Setelah itu, baris p.var_name digunakan untuk menampilkan nama variabel acak yang dimiliki objek tersebut. Karena pada inisialisasi tidak diberikan nama variabel, maka nilai default yang ditampilkan adalah tanda tanya (’?’) yang berasal dari parameter var_name di konstruktor.

In [10]:
(p['low'], p['medium'], p['high'])

(0.125, 0.375, 0.5)

Potongan kode ini digunakan untuk menampilkan nilai probabilitas dari masing-masing kategori yaitu ‘low’, ‘medium’, dan ‘high’ pada objek distribusi probabilitas p yang telah dibuat sebelumnya. Karena frekuensi awal yang dimasukkan adalah 125, 375, dan 500, maka setelah dilakukan normalisasi totalnya menjadi 1000, dan masing-masing nilai probabilitas adalah P(low) = 125/1000 = 0.125, P(medium) = 375/1000 = 0.375, dan P(high) = 500/1000 = 0.5. Dengan demikian, ketika kode (p['low'], p['medium'], p['high']) dijalankan, hasil yang ditampilkan adalah (0.125, 0.375, 0.5).


In [11]:
p.values

['low', 'medium', 'high']

Potongan kode ini digunakan untuk menampilkan daftar nilai atau kategori yang terdapat dalam objek distribusi probabilitas p. Atribut values menyimpan semua kemungkinan nilai dari variabel acak yang telah dimasukkan, yaitu ‘low’, ‘medium’, dan ‘high’. Dengan demikian, ketika p.values dipanggil, hasil yang ditampilkan adalah daftar ['low', 'medium', 'high'] yang merepresentasikan seluruh nilai yang memiliki probabilitas dalam distribusi tersebut.


In [12]:
p = ProbDist('Y')
p['Cat'] = 50
p['Dog'] = 114
p['Mice'] = 64
(p['Cat'], p['Dog'], p['Mice'])

(50, 114, 64)

In [ ]:
p.normalize()
(p['Cat'], p['Dog'], p['Mice'])

(0.21929824561403508, 0.5, 0.2807017543859649)

Potongan kode ini membuat objek distribusi probabilitas bernama p dengan variabel acak “Y”. Nilai-nilai yang mungkin untuk variabel tersebut adalah “Cat”, “Dog”, dan “Mice”, masing-masing diberi frekuensi awal sebesar 50, 114, dan 64. Pada tahap pertama, ketika (p['Cat'], p['Dog'], p['Mice']) dipanggil, hasil yang ditampilkan masih berupa frekuensi mentah, bukan probabilitas, yaitu (50, 114, 64). Setelah itu, fungsi normalize() dijalankan untuk menormalkan total frekuensi menjadi 1. Proses ini dilakukan dengan membagi setiap frekuensi dengan total keseluruhan (50 + 114 + 64 = 228). Dengan demikian, nilai probabilitas hasil normalisasi adalah P(Cat) = 50/228 ≈ 0.219, P(Dog) = 114/228 ≈ 0.5, dan P(Mice) = 64/228 ≈ 0.281. Ketika (p['Cat'], p['Dog'], p['Mice']) dipanggil kembali setelah normalisasi, hasil yang ditampilkan mencerminkan probabilitas tersebut.


In [14]:
p.show_approx()

'Cat: 0.219, Dog: 0.5, Mice: 0.281'

Potongan kode ini digunakan untuk menampilkan isi dari distribusi probabilitas p dalam format teks yang sudah dibulatkan dan disusun berdasarkan urutan nama kategorinya. Fungsi show_approx() mempermudah pembacaan hasil karena menampilkan setiap nilai beserta probabilitasnya dalam satu baris yang rapi. Berdasarkan hasil normalisasi sebelumnya, output yang ditampilkan adalah “Cat: 0.219, Dog: 0.5, Mice: 0.281”, yang menunjukkan bahwa peluang terbesar terdapat pada kategori “Dog” dengan probabilitas 0.5.


## Joint Probability Distribution

The helper function **event_values** returns a tuple of the values of variables in event. An event is specified by a dict where the keys are the names of variables and the corresponding values are the value of the variable. Variables are specified with a list. The ordering of the returned tuple is same as those of the variables.


Alternatively if the event is specified by a list or tuple of equal length of the variables. Then the events tuple is returned as it is.

In [15]:
event = {'A': 10, 'B': 9, 'C': 8}
variables = ['C', 'A']
event_values(event, variables)

(8, 10)

_A probability model is completely determined by the joint distribution for all of the random variables._ (**Section 13.3**) The probability module implements these as the class **JointProbDist** which inherits from the **ProbDist** class. This class specifies a discrete probability distribute over a set of variables. 

In [16]:
psource(JointProbDist)

Values for a Joint Distribution is a an ordered tuple in which each item corresponds to the value associate with a particular variable. For Joint Distribution of X, Y where X, Y take integer values this can be something like (18, 19).

To specify a Joint distribution we first need an ordered list of variables.

In [17]:
variables = ['X', 'Y']
j = JointProbDist(variables)
j

P(['X', 'Y'])

Like the **ProbDist** class **JointProbDist** also employes magic methods to assign probability to different values.
The probability can be assigned in either of the two formats for all possible values of the distribution. The **event_values** call inside  **_ _getitem_ _**  and **_ _setitem_ _** does the required processing to make this work.

In [18]:
j[1,1] = 0.2
j[dict(X=0, Y=1)] = 0.5

(j[1,1], j[0,1])

(0.2, 0.5)

It is also possible to list all the values for a particular variable using the **values** method.

In [19]:
j.values('X')

[1, 0]

## Inference Using Full Joint Distributions

In this section we use Full Joint Distributions to calculate the posterior distribution given some evidence. We represent evidence by using a python dictionary with variables as dict keys and dict values representing the values.

This is illustrated in **Section 13.3** of the book. The functions **enumerate_joint** and **enumerate_joint_ask** implement this functionality. Under the hood they implement **Equation 13.9** from the book.

$$\textbf{P}(X | \textbf{e}) = \alpha \textbf{P}(X, \textbf{e}) = \alpha \sum_{y} \textbf{P}(X, \textbf{e}, \textbf{y})$$

Here **α** is the normalizing factor. **X** is our query variable and **e** is the evidence. According to the equation we enumerate on the remaining variables **y** (not in evidence or query variable) i.e. all possible combinations of **y**

We will be using the same example as the book. Let us create the full joint distribution from **Figure 13.3**.  

In [20]:
full_joint = JointProbDist(['Cavity', 'Toothache', 'Catch'])
full_joint[dict(Cavity=True, Toothache=True, Catch=True)] = 0.108
full_joint[dict(Cavity=True, Toothache=True, Catch=False)] = 0.012
full_joint[dict(Cavity=True, Toothache=False, Catch=True)] = 0.016
full_joint[dict(Cavity=True, Toothache=False, Catch=False)] = 0.064
full_joint[dict(Cavity=False, Toothache=True, Catch=True)] = 0.072
full_joint[dict(Cavity=False, Toothache=False, Catch=True)] = 0.144
full_joint[dict(Cavity=False, Toothache=True, Catch=False)] = 0.008
full_joint[dict(Cavity=False, Toothache=False, Catch=False)] = 0.576

Let us now look at the **enumerate_joint** function returns the sum of those entries in P consistent with e,provided variables is P's remaining variables (the ones not in e). Here, P refers to the full joint distribution. The function uses a recursive call in its implementation. The first parameter **variables** refers to remaining variables. The function in each recursive call keeps on variable constant while varying others.

In [21]:
psource(enumerate_joint)

Let us assume we want to find **P(Toothache=True)**. This can be obtained by marginalization (**Equation 13.6**). We can use **enumerate_joint** to solve for this by taking Toothache=True as our evidence. **enumerate_joint** will return the sum of probabilities consistent with evidence i.e. Marginal Probability.

In [22]:
evidence = dict(Toothache=True)
variables = ['Cavity', 'Catch'] # variables not part of evidence
ans1 = enumerate_joint(variables, evidence, full_joint)
ans1

0.19999999999999998

You can verify the result from our definition of the full joint distribution. We can use the same function to find more complex probabilities like **P(Cavity=True and Toothache=True)** 

In [23]:
evidence = dict(Cavity=True, Toothache=True)
variables = ['Catch'] # variables not part of evidence
ans2 = enumerate_joint(variables, evidence, full_joint)
ans2

0.12

Being able to find sum of probabilities satisfying given evidence allows us to compute conditional probabilities like **P(Cavity=True | Toothache=True)** as we can rewrite this as $$P(Cavity=True | Toothache = True) = \frac{P(Cavity=True \ and \ Toothache=True)}{P(Toothache=True)}$$

We have already calculated both the numerator and denominator.

In [24]:
ans2/ans1

0.6

We might be interested in the probability distribution of a particular variable conditioned on some evidence. This can involve doing calculations like above for each possible value of the variable. This has been implemented slightly differently  using normalization in the function **enumerate_joint_ask** which returns a probability distribution over the values of the variable **X**, given the {var:val} observations **e**, in the **JointProbDist P**. The implementation of this function calls **enumerate_joint** for each value of the query variable and passes **extended evidence** with the new evidence having **X = x<sub>i</sub>**. This is followed by normalization of the obtained distribution.

In [25]:
psource(enumerate_joint_ask)

Let us find **P(Cavity | Toothache=True)** using **enumerate_joint_ask**.

In [26]:
query_variable = 'Cavity'
evidence = dict(Toothache=True)
ans = enumerate_joint_ask(query_variable, evidence, full_joint)
(ans[True], ans[False])

(0.6, 0.39999999999999997)